### PR TITLE
Add github_branch_protection

### PR DESCRIPTION
**Pull request by eliecharra**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #249 
| ❓ Documentation  | yes

## Description

⚠️ To be able to alert when use user `github_branch_protection_v3`, we should create a resource, a deserializer and a finally a middleware to check if there is any `github_branch_protection_v3` resource in state just to send an alert and remove theses resources. Do we really want to do that ? Maybe a line in LIMITATIONS.md may be enough.

⚠️ It seems that the terraform provider is not returning the repository_id, so I've ignored it from the diff

Attentive review required please @moadibfr @wbeuil

### CODE REVIEW

File: pkg/remote/github/github_branch_protection_supplier.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		s.runner.Run(func() (cty.Value, error) {
+			completeResource, err := s.reader.ReadResource(terraform.ReadResourceArgs{
+				Ty: resourcegithub.GithubBranchProtectionResourceType,
+				ID: id,
+			})
+			if err != nil {
+				logrus.Warnf("Error reading %s[%s]: %+v", id, resourcegithub.GithubBranchProtectionResourceType, err)
+				return cty.NilVal, err
+			}
+			return *completeResource, nil
```
Comment (lotoussa) [inline pkg/remote/github/github_branch_protection_supplier.go:49]: You could create a method for this like readProtection ?

File: pkg/resource/github/github _branch_protection_ext.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+			decodedId, err := base64.StdEncoding.DecodeString(*r.RepositoryId)
+			if err == nil {
+				repoId = string(decodedId)
+			}
+		}
+
+		if repoId == "" {
+			return fmt.Sprintf("Branch: %s (Id: %s)", *r.Pattern, r.Id)
+		}
+		return fmt.Sprintf("Branch: %s, RepoId: %s", *r.Pattern, repoId)
```
Comment (lotoussa) [inline pkg/resource/github/github _branch_protection_ext.go:23]: Isn't missing parenthesis for RepoId ?

**Review by lotoussa — COMMENTED** [review]
~LGTM

**Comment by moadibfr** [comment]
> warning To be able to alert when use user github_branch_protection_v3, we should create a resource, a deserializer and a finally a middleware to check if there is any github_branch_protection_v3 resource in state just to send an alert and remove theses resources. Do we really want to do that ? Maybe a line in LIMITATIONS.md may be enough.

Maybe we could add a warning when we find unmanaged repository only ? Saying that this might be a false positive as we don't support v3 ?

**Review by wbeuil — CHANGES_REQUESTED** [review]

File: doc/LIMITATIONS.md [hunk 3]
```diff
@@ -18,3 +18,14 @@
...
+### Github
+- github_branch_protection_v3:
+  - This resource is not supported and will probably never be as it overlap with github_branch_protection.
+    `github_branch_protection` is more suitable for performance purpose.
+    We cannot support these two resources as we don't have any way to discriminate them by enumerating resources from
+    remote side. They represent the same notion but using REST vs GraphQL APIs. Driftctl team recommend you tu use the
+    newer `github_branch_protection`. Or at least ignore all your `github_branch_protection_v3` in driftignore.
+- github_branch_protection:
+  - Branch protection are not returned as unmanaged if the branch protection pattern does not match at least one branch
+  - We cannot show the repository for unmanaged resources as the terraform provider does not return the information
```
[the remark this answers is not in the document]
Comment (eliecharra) [reply doc/LIMITATIONS.md:31]: It wasn't clear when I read my own sentence so I changed to ` We cannot show the related repository name in driftctl output as the terraform provider does not retrieve this information.`

**Comment by eliecharra** [comment]
> Maybe we could add a warning when we find unmanaged repository only ? Saying that this might be a false positive as we don't support v3 ?

We cannot make the difference between a false positive drift and a real drift with this approach. 
For me we the only way to handle this is to read v3 res from the state and try to match them to non-v3 unmanaged one and ignore them with a warning. So we'll end up with a v3 resource and deserializer without any supplier related.

WDYT @moadibfr @wbeuil ? Is it worth it ? I can totally do this in a couple of hours.

File: doc/LIMITATIONS.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wbeuil) [reply doc/LIMITATIONS.md]: I'm fine with that !

**Comment by wbeuil** [comment]
I'm not sure it's worth your time to do it. Are we sure that we can create a non-v3 resource with a v3 resource ? Can't we just warn the user if we find a v3 resource in its state ?

**Review by wbeuil — DISMISSED** [review]

**Review by moadibfr — DISMISSED** [review]

File: pkg/remote/github/github_branch_protection_supplier.go (hunk 1 above)
Comment (eliecharra) [reply pkg/remote/github/github_branch_protection_supplier.go:49]: This part of the code is never reused and it is readable enough, so for me there is no reason to create a new function.

File: pkg/resource/github/github _branch_protection_ext.go (hunk 2 above)
Comment (eliecharra) [reply pkg/resource/github/github _branch_protection_ext.go:23]: This is a good question, WDYT @moadibfr @wbeuil ? Should we display `Branch: %s (RepoId: %s)` to be more consistent with `Branch: %s (Id: %s)` ?

Comment (wbeuil) [reply pkg/resource/github/github _branch_protection_ext.go:23]: Let's be consistent then. But I'm fine with one or the other

Comment (eliecharra) [reply pkg/resource/github/github _branch_protection_ext.go]: Fixed

**Review by wbeuil — APPROVED** [review]

**Pull request merged**